File: golang/2799_count_complete_subarrays/count_complete_subarrays.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func countCompleteSubarrays(nums []int) int {
	res := 0
	cnt := make(map[int]int)
	n := len(nums)
	right := 0
	distinct := make(map[int]bool)
	for _, num := range nums {
		distinct[num] = true
	}
	distinctCount := len(distinct)
	for left := 0; left < n; left++ {
		if left > 0 {
			remove := nums[left-1]
			cnt[remove]--
			if cnt[remove] == 0 {
				delete(cnt, remove)
			}
		}
		for right < n && len(cnt) < distinctCount {
			add := nums[right]
			cnt[add]++
			right++
		}
		if len(cnt) == distinctCount {
			res += (n - right + 1)
		}
	}
	return res
}

func main() {
	tests := []struct {
		nums []int
		ans  int
	}{
		{[]int{1, 3, 1, 2, 2}, 4},
		{[]int{5, 5, 5, 5}, 10},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, countCompleteSubarrays(test.nums), index)
	}
}

File: cpp/2799_count_complete_subarrays/count_complete_subarrays.cpp
#include <cassert>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  int countCompleteSubarrays(vector<int>& nums) {
    int res = 0;
    unordered_map<int, int> cnt;
    int n = nums.size();
    int right = 0;
    unordered_set<int> distinct(nums.begin(), nums.end());
    int distinct_count = distinct.size();

    for (int left = 0; left < n; left++) {
      if (left > 0) {
        int remove = nums[left - 1];
        cnt[remove]--;
        if (cnt[remove] == 0) {
          cnt.erase(remove);
        }
      }
      while (right < n && cnt.size() < distinct_count) {
        int add = nums[right];
        cnt[add]++;
        right++;
      }
      if (cnt.size() == distinct_count) {
        res += (n - right + 1);
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, int>> tests{
      {{1, 3, 1, 2, 2}, 4},
      {{5, 5, 5, 5}, 10},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().countCompleteSubarrays(nums) == ans);
  }
}